### PR TITLE
Auto 2780 sync rs to v1.5.14

### PR DESCRIPTION
**Pull request by nachovizzo**
JIRA : https://dexory.atlassian.net/browse/AUTO-2780, brother PR in https://github.com/botsandus/auto-rslidar-sdk/pull/6

# Upstream diff as baseline

https://github.com/RoboSense-LiDAR/rs_driver/compare/v1.5.10...v1.5.14

# Bump RS driver (ROS independent) from v1.5.10 to v1.5.14

### Added
- Support multiple lidars with different multicast addresses and the same port.
### Fixed
- Fixed the bug that only one lidar was parsed correctly when multiple bp4.0 were used.

### Added
- Support RSMX.
### Fixed
- Update timestamp parsing unit and the number of packets per frame in decoder_RSE1.
- Update firing_tss of Helios/Helios16P/RubyPlus.
- Fix compilation bug of unit test.

### Fixed
- Fix bug in getting device info and status.
- Fix bug in getting device temperature.

### Added
- Split frame at the end of frame.

### Changed
- Enable modify socket buffer size.

### Fixed
- Fix the bug of rs_drvier_viewer.cpp abnormal crash.
- Fix fir_tss of RSBPV4.
- Fix rotation direction of RSBPV4 when reversal.

### CODE REVIEW

File: src/rs_driver/driver/decoder/section.hpp [hunk 1]
```diff
@@ -67,6 +67,13 @@ class AzimuthSection
 #ifndef UNIT_TEST
 private:
 #endif
+
+  static
+  int32_t _round(int32_t value)
+  {
+      return (value % 36000 + 36000) % 36000;
```
Comment (nachovizzo) [inline src/rs_driver/driver/decoder/section.hpp:74]: I get scared every time I learn how these changes went into production...

Comment (richardw347) [reply src/rs_driver/driver/decoder/section.hpp:74]: :scream: :scream_cat: :scream: :scream_cat: :scream: :scream_cat:

File: CMakeLists.txt [hunk 2]
```diff
@@ -6,7 +6,7 @@ if(WIN32)
   cmake_policy(SET CMP0074 NEW) # CMake 3.12 required
 endif(WIN32)
 
-project(rs_driver VERSION 1.5.10)
+project(rs_driver VERSION 1.5.13)
```
Comment (nachovizzo) [inline CMakeLists.txt:9]: Fuck... so many errors!!!!!

File: doc/howto/07_how_to_port_your_app_from_rs_driver_1.3.x_to_1.5.x.md [hunk 3]
```diff
@@ -126,6 +126,7 @@ typedef struct RSInputParam  ///< The LiDAR input parameter
   bool use_vlan = false;                       ///< Vlan on-off
   uint16_t user_layer_bytes = 0;    ///< Bytes of user layer. thers is no user layer if it is 0
   uint16_t tail_layer_bytes = 0;    ///< Bytes of tail layer. thers is no tail layer if it is 0
+  uint32_t socket_recv_buf = 106496;   //<Bytes of socket receive buffer.
```
Comment (nachovizzo) [inline doc/howto/07_how_to_port_your_app_from_rs_driver_1.3.x_to_1.5.x.md:129]: Just an unused parameter if we don't enable this functionality, so safe thing here

File: src/rs_driver/driver/decoder/basic_attr.hpp [hunk 4]
```diff
@@ -119,7 +119,30 @@ inline uint64_t parseTimeUTCWithUs(const RSTimestampUTC* tsUtc)
 
   return (sec * 1000000 + us);
 }
+inline uint64_t parseTimeUTCWithNs(const RSTimestampUTC* tsUtc)
```
Comment (nachovizzo) [inline src/rs_driver/driver/decoder/basic_attr.hpp:122]: Unused function, thanks RS

File: src/rs_driver/driver/decoder/decoder.hpp [hunk 5]
```diff
@@ -351,22 +353,36 @@ inline void Decoder<T_PointCloud>::enableWritePktTs(bool value)
 }
 
 template <typename T_PointCloud>
-inline float Decoder<T_PointCloud>::getTemperature()
+inline bool Decoder<T_PointCloud>::getTemperature(float& temp)
```
Comment (nachovizzo) [inline src/rs_driver/driver/decoder/decoder.hpp:356]: @richardw347 they apparently now implemented how to get the temperature directly from the driver now, FYI

Once again, this function is not being called anywhere I can see, so it's just implemented and we will need to call it from the ROS driver to actually fetch the temperature

File: src/rs_driver/driver/decoder/decoder.hpp [hunk 6]
```diff
@@ -351,22 +353,36 @@ inline void Decoder<T_PointCloud>::enableWritePktTs(bool value)
...
+  {
+    return false;
+  }
+
+  temp = temperature_;
+  return true;
 }
 
 template <typename T_PointCloud>
 inline bool Decoder<T_PointCloud>::getDeviceInfo(DeviceInfo& info)
```
Comment (nachovizzo) [inline src/rs_driver/driver/decoder/decoder.hpp:368]: As mentioned in https://github.com/botsandus/auto-rslidar-sdk/pull/6#discussion_r1680725176 this is now implemented but we don't have free access to  it unless we put a call to this function in the ROS driver

File: src/rs_driver/driver/decoder/decoder.hpp [hunk 7]
```diff
@@ -351,22 +353,36 @@ inline void Decoder<T_PointCloud>::enableWritePktTs(bool value)
...
+  }
+  info = device_info_;
   return true;
 }
 
 template <typename T_PointCloud>
 inline bool Decoder<T_PointCloud>::getDeviceStatus(DeviceStatus& status)
 {
-  memcpy (&status, &device_status_, sizeof(DeviceStatus));
+  if(!device_status_.state)
```
Comment (nachovizzo) [inline src/rs_driver/driver/decoder/decoder.hpp:381]: Same here, we don't call this function. Check https://github.com/botsandus/auto-rslidar-sdk/pull/6#discussion_r1680725176

File: src/rs_driver/driver/decoder/decoder_RSHELIOS.hpp [hunk 8]
```diff
@@ -145,10 +145,10 @@ inline RSDecoderMechConstParam& DecoderRSHELIOS<T_PointCloud>::getConstParam()
   float blk_ts = 55.56f;
   float firing_tss[] = 
   {
-    0.00f,  1.57f,  3.15f,  4.72f,  6.30f,  7.87f,  9.45f, 11.36f, 
-    13.26f, 15.17f, 17.08f, 18.99f, 20.56f, 22.14f, 23.71f, 25.29f,
-    26.53f, 29.01f, 27.77f, 30.25f, 31.49f, 33.98f, 32.73f, 35.22f, 
-    36.46f, 37.70f, 38.94f, 40.18f, 41.42f, 42.67f, 43.91f, 45.15f
+    0.00f,  1.73f,  3.46f,  5.19f,  6.92f,  8.65f,  10.38f, 12.11f,
```
Comment (nachovizzo) [inline src/rs_driver/driver/decoder/decoder_RSHELIOS.hpp:148]: @doisyg  they also updated the firing timestamps for Helios. We still don't deskew, but we should suspect about these numbers once we decide to deskew that lidar

File: src/rs_driver/driver/input/unix/input_sock_select.hpp [hunk 9]
```diff
@@ -189,13 +193,19 @@ inline int InputSock::createSocket(uint16_t port, const std::string& hostIp, con
     }
   }
 
-#ifdef ENABLE_DOUBLE_RCVBUF
+#ifdef ENABLE_MODIFY_RECVBUF
   {
-    uint32_t opt_val;
+    uint32_t opt_val = input_param_.socket_recv_buf, before_set_val,after_set_val = 0;
```
Comment (nachovizzo) [inline src/rs_driver/driver/input/unix/input_sock_select.hpp:198]: these lines are connected to the discussion from https://github.com/botsandus/auto-rslidar-sdk/pull/6#discussion_r1680737847

File: src/rs_driver/driver/lidar_driver_impl.hpp [hunk 10]
```diff
@@ -405,10 +403,6 @@ void LidarDriverImpl<T_PointCloud>::splitFrame(uint16_t height, double ts)
     cb_put_cloud_(cloud);
     decoder_ptr_->point_cloud_ = getPointCloud();
   }
-  else
```
Comment (nachovizzo) [inline src/rs_driver/driver/lidar_driver_impl.hpp:408]: I have no idea why they removed this. Let's believe it's a safe change :crossed_fingers:

File: src/rs_driver/driver/decoder/decoder_RSBP.hpp [hunk 11]
```diff
@@ -274,9 +303,12 @@ inline bool DecoderRSBP<T_PointCloud>::internDecodeMsopPkt(const uint8_t* packet
 
       int32_t angle_vert = this->chan_angles_.vertAdjust(chan);
       int32_t angle_horiz_final = this->chan_angles_.horizAdjust(chan, angle_horiz);
-
       float distance = ntohs(channel.distance) * this->const_param_.DISTANCE_RES;
-
+      if(reversal_)
```
Comment (nachovizzo) [inline src/rs_driver/driver/decoder/decoder_RSBP.hpp:307]: I'm not sure what changed here, but it looks to be a follow-up from https://github.com/botsandus/auto-rslidar-sdk/pull/3#pullrequestreview-1788792984, which we don't care about, as it seems to be a commissioning parameter we don't set
image

File: src/rs_driver/driver/decoder/decoder_RSBP.hpp [hunk 12]
```diff
@@ -201,23 +210,43 @@ inline bool DecoderRSBP<T_PointCloud>::internDecodeMsopPkt(const uint8_t* packet
...
+    isFirstPkt_ = false;
+    if ((pkt.header.lidar_type == 0x03) && (pkt.header.lidar_model == 0x04)) 
+    {
+      isBpV4_ = true;
+      this->const_param_.DISTANCE_RES = 0.0025f;
+      this->mech_const_param_.RX = 0.01619f;
+      this->mech_const_param_.RY = 0.0085f;
+      this->mech_const_param_.RZ = 0.09571f;
+      float blk_ts = 55.56f;
+      float firing_tss[] =
```
Comment (nachovizzo) [inline src/rs_driver/driver/decoder/decoder_RSBP.hpp:227]: lol, they changed this after we complained in GH and also by email https://github.com/RoboSense-LiDAR/rs_driver/issues/41

 sanity check here, we have the same numbers in our launch files https://github.com/botsandus/auto-robot/blob/f11b523199fb4231a8aa842ade0fb998a16bfd7a/dexory_common/dexory_common/launch/bpearl_interface.py#L52-L85
```py
  _BEAM_FIRING_TIMES_V4 = [
        0.00, 1.67, 3.34, 5.00, 6.67, 8.34, 10.01, 11.68,
        13.34, 15.01, 16.68, 18.35, 20.02, 21.68, 23.35, 25.02,
        26.69, 28.36, 30.02, 31.69, 33.36, 35.03, 36.70, 38.36,
        40.03, 41.70, 43.37, 45.04, 46.70, 48.37, 50.04, 51.71,
    ]
```

**Review by nachovizzo — COMMENTED** [review]
Besides all the comments, I don't see any potential issue. So whenever we decide for the socket size change we are ready to test and merge

File: src/rs_driver/driver/decoder/decoder_RSBP.hpp (hunk 12 above)
Comment (doisyg) [reply src/rs_driver/driver/decoder/decoder_RSBP.hpp:227]: how beautiful

File: src/rs_driver/driver/decoder/decoder.hpp (hunk 5 above)
Comment (richardw347) [reply src/rs_driver/driver/decoder/decoder.hpp:356]: Thanks for the heads up it's worth testing but will probably stick with the web scraping till we get the rest of the metrics

**Comment by nachovizzo** [comment]
Tested on 5 different robots, did all sorts of checks, CPU/mem consumption in the last 30 days. Data grabs + mapping, obstacle detection, ros2 topic hz, delay, etc, etc and could not brake the new driver so it's good to go for me

https://github.com/botsandus/auto/pull/421

**Review by doisyg — APPROVED** [review]

**Pull request merged**